File: src/brechabot/scoring.py
from __future__ import annotations

from .models import Finding, Verdict

_GRADE_BANDS = ((90, "A"), (80, "B"), (70, "C"), (60, "D"))


def grade_for(score: int) -> str:
    for cutoff, grade in _GRADE_BANDS:
        if score >= cutoff:
            return grade
    return "F"


def summarize(findings: list[Finding]) -> dict:
    total = len(findings)
    vulnerable = sum(1 for f in findings if f.verdict is Verdict.VULNERABLE)
    partial = sum(1 for f in findings if f.verdict is Verdict.PARTIAL)
    safe = sum(1 for f in findings if f.verdict is Verdict.SAFE)
    unknown = total - vulnerable - partial - safe
    score = round(((safe + partial * 0.5) / total) * 100) if total else 0

    by_category: dict[str, dict[str, int]] = {}
    by_owasp: dict[str, dict[str, int]] = {}
    for f in findings:
        cat = by_category.setdefault(f.category, {"total": 0, "vulnerable": 0, "partial": 0})
        cat["total"] += 1
        if f.verdict is Verdict.VULNERABLE:
            cat["vulnerable"] += 1
        if f.verdict is Verdict.PARTIAL:
            cat["partial"] += 1
        owasp_key = f.owasp or "Unmapped"
        owasp = by_owasp.setdefault(owasp_key, {"total": 0, "vulnerable": 0})
        owasp["total"] += 1
        if f.verdict is Verdict.VULNERABLE:
            owasp["vulnerable"] += 1

    return {
        "total": total,
        "vulnerable": vulnerable,
        "partial": partial,
        "safe": safe,
        "unknown": unknown,
        "score": score,
        "grade": grade_for(score),
        "byCategory": by_category,
        "byOwasp": by_owasp,
    }

File: src/scoring.ts
import type { CategoryRollup, Finding, OwaspRollup, Summary } from "./models.js";

const GRADE_BANDS: [number, string][] = [[90, "A"], [80, "B"], [70, "C"], [60, "D"]];

export function gradeFor(score: number): string {
  for (const [cutoff, grade] of GRADE_BANDS) if (score >= cutoff) return grade;
  return "F";
}

// Python round() is banker's rounding (round-half-to-even). Match it.
function roundHalfEven(x: number): number {
  const floor = Math.floor(x);
  const diff = x - floor;
  if (diff < 0.5) return floor;
  if (diff > 0.5) return floor + 1;
  return floor % 2 === 0 ? floor : floor + 1;
}

export function summarize(findings: Finding[]): Summary {
  const total = findings.length;
  const vulnerable = findings.filter((f) => f.verdict === "VULNERABLE").length;
  const partial = findings.filter((f) => f.verdict === "PARTIAL").length;
  const safe = findings.filter((f) => f.verdict === "SAFE").length;
  const unknown = total - vulnerable - partial - safe;
  const score = total ? roundHalfEven(((safe + partial * 0.5) / total) * 100) : 0;

  const byCategory: Record<string, CategoryRollup> = {};
  const byOwasp: Record<string, OwaspRollup> = {};
  for (const f of findings) {
    const cat = (byCategory[f.category] ??= { total: 0, vulnerable: 0, partial: 0 });
    cat.total += 1;
    if (f.verdict === "VULNERABLE") cat.vulnerable += 1;
    if (f.verdict === "PARTIAL") cat.partial += 1;
    const key = f.owasp || "Unmapped";
    const ow = (byOwasp[key] ??= { total: 0, vulnerable: 0 });
    ow.total += 1;
    if (f.verdict === "VULNERABLE") ow.vulnerable += 1;
  }

  return { total, vulnerable, partial, safe, unknown, score, grade: gradeFor(score), byCategory, byOwasp };
}
